### PR TITLE
Appbundle

### LINKED ISSUE #330: Build app as appbundle
Google considers it to be best practice to build an app as appbundle and not APK.

Should be tested in Firebase.

### PR DESCRIPTION
**Pull request by fremartini**
Closes #330

### CODE REVIEW

**Review by marfavi — CHANGES_REQUESTED** [review]
Fix merge conflicts

**Comment by jonasanker** [comment]
@fremartini There has been a larger refactoring in #517 which you should take into account. It will probably give you a few bumps :)

Secondly. I would be mindful of not over-engineering with action files. I can see the value where several workflows have a need to call an action. In cases where an action is only called by one function, it might be just fine to have the steps in the workflow file :)

Thirdly, I am not sure but we should verify that Github Action supports parsing secrets as inputs such that we not leak secret values by accident when parsed as a input.

**Review by jonasanker — CHANGES_REQUESTED** [review]
We should transition to building app bundles instead of apk's. We might need to update Firebase to support app bundle distribution with [Firebase App distribution](https://firebase.google.com/codelabs/appdistribution-app-bundles#0)

File: .github/actions/build_android/action.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+runs:
+  using: "composite"
+  steps:
+    - name: Setup Flutter environment
+      uses: ./.github/actions/setup_flutter_environment
+      with:
+        prod_release: ${{ inputs.prod_release }}
+
+    - name: Build apk (dev)
```
Comment (marfavi) [inline .github/actions/build_android/action.yml:25]: Update name to reflect an app bundle being built

**Review by marfavi — CHANGES_REQUESTED** [review]

**Comment by jonasanker** [comment]
@fremartini Can you resolve the merge conflicts and then I'll provide my review :)

**Review by marfavi — COMMENTED** [review]
I will leave this up to @jonasanker to review

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-name: Build and test
+name: Build application
```
Comment (jonasanker) [inline .github/workflows/build.yml:1]: I would prefer that the build step also invoke the test steps because they are so closely related

File: .github/workflows/release-dev.yml [hunk 3]
```diff
@@ -5,17 +5,17 @@ on:
...
-      prodRelease: false
+      store_artifacts: true
+      prod_release: false
     secrets: inherit
 
-  dev_upload_ios:
+  upload_ios:
     name: Upload iOS build to Firebase App Distribution
     runs-on: ubuntu-latest
-    needs: [build_and_test]
```
Comment (jonasanker) [inline .github/workflows/release-dev.yml:18]: Should we release the app if tests are failing? There might be a slight argument for Dev but I would not accept releasing the app on Prod if tests were failing

File: .github/workflows/release-prod.yml [hunk 4]
```diff
@@ -2,21 +2,21 @@ name: Release Production App
...
+      store_artifacts: true
+      prod_release: true
+      tag_name: ${{ github.event.release.tag_name }}
     secrets: inherit
 
-  prod_release_ios:
+  release_ios:
     name: Upload iOS build to App Store connect
     runs-on: macos-latest
-    needs: [build_and_test]
```
Comment (jonasanker) [inline .github/workflows/release-prod.yml:19]: We should not release the Production App if tests are failing, thus we need a dependency on tests

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+name: Test
```
Comment (jonasanker) [inline .github/workflows/test.yml:1]: Possibly we will include tests as part of the build step (my argument). If we end up different, we must also run the tests pipeline on the main branch for coverage reports to function

File: .github/actions/build_ios/action.yml [hunk 6]
```diff
@@ -0,0 +1,64 @@
+name: Build iOS app
+description: Build and upload iOS app
+inputs:
```
Comment (jonasanker) [inline .github/actions/build_ios/action.yml:3]: We should verify that secret values are not exposed when they are parsed as inputs

File: .github/actions/build_ios/action.yml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  version:
+    required: true
+    description: "App version"
+  build_version:
+      required: true
+      description: "Build version"
+  apple_ios_signing_cert:
+      required: true
+      description: "iOS signing certificate"
```
Comment (jonasanker) [inline .github/actions/build_ios/action.yml:12]: Add .. in base64?

**Review by jonasanker — CHANGES_REQUESTED** [review]
Cool refactoring. I have added some comments

File: .github/actions/build_ios/action.yml (hunk 6 above)
Comment (fremartini) [reply .github/actions/build_ios/action.yml:3]: Github Actions redacts secrets automatically regardless of how they are passed around. From the logs:

2023-10-29T10:31:45.9020320Z ##[group]Run .github/scripts/setup_certs.sh
2023-10-29T10:31:45.9021020Z [36;1m.github/scripts/setup_certs.sh[0m
2023-10-29T10:31:45.9156510Z shell: /bin/bash --noprofile --norc -e -o pipefail {0}
2023-10-29T10:31:45.9157150Z env:
2023-10-29T10:31:45.9157490Z   PROD_RELEASE: false Environment variable
2023-10-29T10:31:45.9157890Z   TAG_NAME: dev <-- Environment variable
2023-10-29T10:31:45.9204640Z   APPLE_IOS_SIGNING_CERT: *** <-- Secret

2023-10-29T10:31:45.9205180Z   APPLE_IOS_SIGNING_CERT_PW: *** <-- Secret
2023-10-29T10:31:46.8769810Z   APPLE_IOS_PROVISIONING_PROFILE: *** <-- Secret

2023-10-29T10:31:46.8770530Z   APPLE_KEYCHAIN_PW: *** <-- Secret

File: .github/workflows/release-dev.yml (hunk 3 above)
Comment (fremartini) [reply .github/workflows/release-dev.yml:18]: No the app should not be released if tests are failing. Pull requests should not (and cannot?) be merged into main if tests fail. We will therefore never have a build failing the test suite on main

File: .github/workflows/release-prod.yml (hunk 4 above)
Comment (fremartini) [reply .github/workflows/release-prod.yml:19]: For the same reason above i do not believe this will be relevant

File: .github/workflows/build.yml (hunk 2 above)
Comment (fremartini) [reply .github/workflows/build.yml:1]: I would prefer if the 'build' actions sole responsibility is to build the app and have a separate job for testing the app. We will not be able to merge pull requests into main if any of the pipelines fail either way. I believe this improves the readability of the actions and makes future refactoring easier.

File: .github/actions/build_ios/action.yml (hunk 6 above)
Comment (jonasanker) [reply .github/actions/build_ios/action.yml:3]: <3

**Review by jonasanker — CHANGES_REQUESTED** [review]
Tests must be run on the main branch to be able to make a code coverage diff for other branches.

File: .github/workflows/build.yml (hunk 2 above)
Comment (jonasanker) [reply .github/workflows/build.yml:1]: For me, it's an edge case that Flutter supports running tests without an explicit build. 

It would have been the ideal case that tests were run on one of the build artefacts we create. Now the risk is that an app could possibly be built differently than the variant where tests are run, thus this could lead to functionality not working on a particular build.

Anyhow the argument of separation is fine. Would it then make sense to have a wrapper workflow which calls both build and test?

File: .github/workflows/release-dev.yml (hunk 3 above)
Comment (jonasanker) [reply .github/workflows/release-dev.yml:18]: My concern is that this principle is not guarded by pipelines but by a particular configuration we have for our branch protection. I think the contract is more clear when the dependency is expressed in the pipeline

File: .github/workflows/build.yml (hunk 2 above)
Comment (fremartini) [reply .github/workflows/build.yml:1]: I'm not sure i see how the test build could differ from the actual build? The code is checked out and generated using the 'setup flutter environment' action for both builds. Anyhow i have made the test workflow run on pushes to the main branch

Comment (jonasanker) [reply .github/workflows/build.yml:1]: We are building a specific Android and iOS variant. We don't have a test case for it yet but we already have functionality which are native to the build, thus a test case could potentially behave differently on iOS vs Android.

File: .github/workflows/release-dev.yml (hunk 3 above)
Comment (fremartini) [reply .github/workflows/release-dev.yml:18]: I understand your concern but i don't see why you would be able to complete a PR with a failing pipeline to start with? That surely indicates something is wrong

Comment (jonasanker) [reply .github/workflows/release-dev.yml:18]: The tests are run on the checked out branch. This might be out of sync with the master branch and thus you can't be sure that it will have the same effect on the main branch unless you enforce the constraint that branch is always in sync with main 🙂

File: .github/workflows/build.yml (hunk 2 above)
Comment (jonasanker) [reply .github/workflows/build.yml:1]: The MobilePay integration reacting to the deep link. That's what my memory tells me. Really hope I'm remembering correctly 🙃

**Review by jonasanker — APPROVED** [review]
Oh the joy of principles. Tak Martini ✅✅✅
